Overhaul for Dewesoft's newest DWDataReader library (5.0.0)

Bojan023 opened the pull request:
Hi all,

I've spent the weekend to overhaul this Python wrapper since Dewesoft recently released a new version of their Data Reader (5.0.0.). All the currently implemented DLL function calls are marked deprecated in version 5.0.0, and new functions are introduced to replace the old ones. The most impactful change is that instantiation is now being handled by Dewesoft's library internally. This saves a couple mechanisms and workarounds in the wrapper, but required to change all function calls to the DLL.

Non-exhaustive list of changes:

- Update DWDataReader DLL to 5.0.0
- New/adapted structs for the new DLL functions
- Introduced logic for a reader_handle for reader instances that must be passed to new DLL functions
- Dropped some logic for keeping track of instances, files being opened, etc. Not required with new DLL.
- Added functionality to extract column names from channel XML for a multimensional array
- Added a couple of classes for proper channel types, binary channels, etc.
- Dropped pytz dependency

Breaking changes:

- Moving forward with requiring Python >3.10 (abc import, | operator, probably few things more). Older Python versions can use the older versions of this module.
- Some internal return type changes to speed things up slightly
- Removed plotting function; reduces dependencies and brings focus to the core functionality of being a wrapper
- Renaming of dw.open() as this can conflict with the standard library (from dwdatareader import *)
- All tests pass except for test_context. As this test was mostly related to the previously required DLL reader instance tracking logic, which is now removed. Therefore, I marked the test as expected to fail.

For my specific application (sync and async full speed data) and configuration it seems to work flawlessly, but it could be tested to some other .dxd configs to buff and polish. All in all I think with these changes we're nicely up-to-date again!

Looking forward to hear your thoughs and see some more testing on different configs.

ajlamarc commented:
@Bojan023 This is great, thanks! I've tested it on a set of multifiles with 30+ GB of data and no issues. It also seems to run faster than before.

I was seeing a bug with the 4.x DLL used by the current release of `dwdatareader`, where it reported a specific file is corrupt, that was fixed with this 5.0 DLL.  I'm not very familiar with Dewesoft, but it seems like some .dxd files produced in newer Dewesoft versions like 2025.1 can be incompatible with previous DLLs ..?

costerwi commented on dwdatareader/__init__.py:
```diff
@@ -1,171 +1,242 @@
+        # Fallback to direct library name loading
+        return loader(library_name)
+
+DLL: ctypes.CDLL | ctypes.WinDLL = load_library()
```
On Linux this raises:
AttributeError: module 'ctypes' has no attribute 'WinDLL'

Bojan023 replied:
https://github.com/costerwi/dwdatareader/pull/71/commits/b5b8383513c0c8d6059000c5a91131062b7ea8bf shiuld fix this

costerwi replied:
This works but since WinDLL is a subclass of CDLL, I believe you may simply do this:
```python
DLL: ctypes.CDLL = load_library()
```

Bojan023 commented on dwdatareader/__init__.py:
```diff
@@ -191,326 +262,353 @@ def offset(self):
-            os.remove(self.name)
-            self.delete = False
+            self.channels = (DWChannel * 0)()  # Delete channel metadata
+            del self
```
Oops, this was a left-over from some experimenting with the Python garbage collector. Fixed in https://github.com/costerwi/dwdatareader/pull/71/commits/b06ad552ed6398a9489cb43c6754f76bbd7c9304

Bojan023 commented:
> @Bojan023 This is great, thanks! I've tested it on a set of multifiles with 30+ GB of data and no issues. It also seems to run faster than before.

Great! Thanks for testing. What OS are you using? Do you have any .dxd (or .d7d) files with reduced channel by any chance?

costerwi reviewed (changes requested):
I think there are a lot of nice ideas here!
I've added a few initial suggestions to simplify and improve things. Please take a look. Thanks!

Bojan023 commented on dwdatareader/__init__.py:
```diff
@@ -287,230 +364,220 @@ def reduced(self):
-            # Return dataframe of ALL channels by default
-            channels = self.keys()
-        return pandas.DataFrame({k: self[k].series() for k in channels})
+    def dataframe(self, channels: List[DWChannel] = None) -> pd.DataFrame:
```
I agree that list of str is preferred from a user friendliness point of view, however this also demands that channels have unique names. I think it is unlikely to have the same channel name for multiple channels, but it is also not unimaginable. Especially if plugins are used, with default column names this will end up in duplicates and will rule out using this module. What's your take on this? We can also move on with list of str and mark this as technical debt to solve later?

Bojan023 replied:
I think that Dewesoft has no logic for the names of channels. The function DLL's explicitly require an index so I assume it is index-based (and thus unique indices) where the name of a channel is a simple attribute. We're taking the opposite approach by indexing on channel name, which index as attribute for conveniance.

fleimgruber replied:
As heavy Power Modules users this has been bugging us as well for a long time, see also my colleagues #34. I would also vote for changing the default behavior of just returning the first matching channel name as that is non-obvious behavior that makes it harder to debug.

Bojan023 replied:
I don't think these are unique either. Here's a channel from Math:
<img width="449" height="214" alt="image" src="https://github.com/user-attachments/assets/d31fbe17-b521-46a6-999b-b3904799c810" />

And one from the NMEA plugin:
<img width="470" height="233" alt="image" src="https://github.com/user-attachments/assets/e8438d6c-addd-476b-99cb-70b0226492d9" />

My conclusion is that not every plugin guarantees a unique longname

fleimgruber replied:
I am not sure. They might be now, but this could potentially change in future DWDataReader DLL upgrades without notice. We should rely on the known unique indexes, keep those as references and give the users ways to disambiguate on the channels using other channel metadata.

costerwi replied:
@Bojan023 your example is showing two unique long_names.. How do you conclude they are not guaranteed unique? That seems to be the intent of the long name described at the end of #34 

@fleimgruber please be specific about which are the "known unique indexes". I'm not clear on what those are right now. Thanks!

Bojan023 replied:
As mentioned by @fleimgruber, the usage of multiple plugins yields channels with the same name. The NMEA plugin in my examples generates a general long_name without any notion of an instance or prefix. Strictly speaking it is speculative but another instance would very likely yield the exact same long_name. This would result in a similar situation as @fleimgruber mention/ with the Power Modules.

The Math plugin seems to add the original channel name as prefix (I named the Math channel the same as the sync channel it is based on). But even with this prefix, two math channels based on the same sync channel will likely yield the same long_name.

fleimgruber replied:
> @fleimgruber please be specific about which are the "known unique indexes". I'm not clear on what those are right now. Thanks!

sorry for the late reply, in the meantime I am able to quote you on this one :) "I agree that we could use channel.index as the key if that is unique" from https://github.com/costerwi/dwdatareader/pull/71#issuecomment-3172562019.

Bojan023 commented:
Small oversight: if using the DWFile dict instance also for channels the keys are not only channels. So I would suggest to re-introduce self.channels as dict and point DWFile __len__, __getattr__ and __iter__ methods to self.channel

<img width="634" height="197" alt="image" src="https://github.com/user-attachments/assets/ef8286fd-a78b-4365-9348-15de31da1440" />

Bojan023 commented:
That is interesing, as my screenshot shows all keys in a DWFile instance, where not only channels are shown as keys. So iterating over this dict will yield channels and attributes. Where is the logic for separating channels from non-channels keys? 

Another problem I encountered is that mixing binary channels and non-binary channels results in a faulty dataframe, especially since multiple binary values can occur at the exact same timestamp. This actually holds for any async channe. I think we should separate sync from async channels and provide separate methods for their respective dataframes.

costerwi commented:
> That is interesing, as my screenshot shows all keys in a DWFile instance, where not only channels are shown as keys. So iterating over this dict will yield channels and attributes. Where is the logic for separating channels from non-channels keys?

I'm not sure what your app is showing you but Python will only iterate over the dict keys, not the attributes or properties. There is nothing required to inherit this behavior. The new tests `test_keys()` and `test_items()` demonstrate this. Also the `len` will only show the number of keys. We should probably add `test_len()`

fleimgruber commented:
I would like to contribute some testing. What is the exact code being executed that shows the conflict your are discussing?

Bojan023 commented:
I have no idea what was wrong this afternoon, somehow my latest commits didn't make it to Github. Anyway, @costerwi suggestions are now implemented and work as intended. Sorry for the confusion. Yet, the channel name keying-issue is not addressed. Any suggestions on how to proceed?

fleimgruber commented:
I see overlap between your proposal and https://github.com/costerwi/dwdatareader/pull/71#discussion_r2254699140. I would agree regarding the default behavior and end-user enablement. I don't have enough experience with sync vs. async channels, but if the use-cases and mechanics are sufficiently similar, why not?

I think the user expectation of being able to directly index the DWFile is still valid and we should provide a way to still do that as an option with guidance on when and how to disambiguate if necessary? Indexing with DWChannel instances won't work in any case because those instances are not valid keys, but maybe I misunderstood you there.

In any case, regarding thoughts and suggestions on how best to proceed: I still stand by my comments about drafting user facing API variants and implementing those as executable minimal examples in testing code.

Bojan023 commented:
> I don't use these features. Can you explain the problem in more detail? What makes the dataframe faulty?

<img width="537" height="320" alt="image" src="https://github.com/user-attachments/assets/bb1338c9-3ead-44f3-8028-b7b496df8858" />

In the example above, G1L-I-X is a sync channel, Velocity and NMEA are async channels. Velocity has a single value per timestamp, however, NMEA log can have multiple values per timestamp. The current DWFile .dataframe() method extracts unique timestamp values (https://github.com/Bojan023/dwdatareader/blob/df3f601e467446a0795d9d3f7ed612120211a478/dwdatareader/__init__.py#L309) which is required for reindexing. For sync channels, this is uniqueness extraction is redundant; sync channels have per definition unique timestamps. For async data, (parts of) the raw data is discarded silently if timestamps are not unique (like the NMEA log channel) and this is not obvious to the user.

I rewrote the .dataframe()-method to get rid of the unique timestamp requirement (using pd.merge and an outer join instead of concatenation). This enables to combine sync and async channels. The example above was generated with this code.

`    def dataframe(self, channels: List = None) -> pd.DataFrame:
        """Return dataframe of selected channels"""
        if channels is None:
            # Return dataframe of all channels by default
            channels = [ch.name for ch in self.values()]

        channel_dfs = [self[ch_name].dataframe() for ch_name in channels]
        for i, ch_df in enumerate(channel_dfs):
            if not i:
                df = ch_df
            else:
                df = pd.merge(df, ch_df, left_on=df.index, right_on=ch_df.index, how='outer')
                df.index = df['key_0'].values
                df = df.drop(columns=['key_0'])
`

The result is a Dataframe with sync and async channels combined, which is a quite complex Dataframe with gaps, many NaN's, etc. In this form, no single column can be used as-is, which I think is not very friendly and error-prone. Explicitly separating sync and async (e.g by sync_dataframe() and async_dataframe() methods) makes it is clear to the user what is being dealt with, and it serves the raw data without any logic (nothing discarded, combined or required to be unique).

> 
> If you already have the DWChannel instance you're looking for then why would you need to index to find it again? `channel = d7d[channel]` doesn't seem useful. I agree that we could use `channel.index` as the key if that is unique. Maybe the `channel.index` could be the key for the DWFile dict and we could add a separate method to search by long name?
> 
> > I still stand by my comments about drafting user facing API variants and implementing those as executable minimal examples in testing code.
> 
> I agree! Can you both comment with some examples of how you would like to obtain channels from a DWFile?

Bojan023 commented:
The rewritten method above makes the dataframe test fail, since there are more rows in the raw data then currently returned. This is due to the non-unique timestamps of async channels, see below.

<img width="739" height="338" alt="image" src="https://github.com/user-attachments/assets/37281592-c8a8-49a9-856f-6f00e65f27eb" />

The first two channels are sync, the latter two async. Arguably not all rows of the async channels are relevant given the single timestamp they represent, but in my view this should be handled by the user instead of discarding all but the first entry silently (as per the current implementation).

Bojan023 commented:
I've implemented separate methods for handling both synchronous and asynchronous channels separatly. In addition, tests have been introduced and updated to reflect these methods. A new ignore_channels argument has been added to the dataframe() methods, allowing exclusion of channels. Can be useful for cases like "give me all channels except...".

Channel identification still relies on unique channel names, and for now, I'm keeping that approach unchanged. The recent overhaul took quite some effort, so I’m pausing further structural changes for the moment. I do plan to add some docstrings (no promises!) and possibly make a few minor refinements, but overall, I believe we're nearing a release-ready state.

From my perspective, the final missing pieces are:

- An updated notebook with some (more) practical examples
- Additional testing with varied measurement files (please confirm that everything works with your files!)

Looking forward to hear any thoughts or feedback!

fleimgruber commented on test_module.py:
```diff
@@ -23,81 +18,85 @@ def setUp(self):
         self.d7dname = os.path.join(os.path.dirname(__file__),
                 "Example_Drive01.d7d")
 
+    # @unittest.expectedFailure
```
Remove dead code in comments.

fleimgruber commented on dwdatareader/__init__.py:
```diff
@@ -1,171 +1,347 @@
+"""Python module that wraps Dewesoft DWDataReaderLib.dll for interactive use with Pyton
 
-@author: shelmore and costerwisch
+@author: shelmore, costerwisch and Bojan023
```
@costerwi should we split this into `maintainers` and `contributors`?

costerwi replied:
```suggestion
Homepage https://github.com/costerwi/dwdatareader
```

fleimgruber commented on dwdatareader/__init__.py:
```diff
@@ -1,171 +1,347 @@
-    """Structure to hold metadata about DWFile"""
+import platform
+import atexit
+import numpy as np
```
Since this PR also contains a fair bit of refactoring, let's take the opportunity to also comply with https://peps.python.org/pep-0008/#imports, i.e. separate stdlib imports from 3rd party imports.

fleimgruber commented on dwdatareader/__init__.py:
```diff
@@ -191,326 +367,425 @@ def offset(self):
+        return pd.DataFrame(data, index=data['time_stamp'],
+                columns=['ave', 'min', 'max', 'rms'])
 
+class DWComplex(ctypes.Structure):
```
This is unused, remove it for now.

fleimgruber reviewed (changes requested):
I did a quick review to get more familiar with the changes for #75.

Bojan023 commented:
Hi @costerwi, can you have a look at the latest commits?

costerwi commented on dwdatareader/__init__.py:
```diff
@@ -520,44 +782,132 @@ def __enter__(self):
+    err_msg = create_string_buffer(err_msg_len.value)
+    err_status = ctypes.c_int(DWStatus.DWSTAT_OK)
+
+    while DLL.DWGetLastStatus(ctypes.byref(err_status), err_msg,
```
Sorry, I still don't understand. This function receives the status of a the previous operation as its only parameter. Why does it need to call `DLL.DWGetLastStatus` and why is that within a `while` loop? Could the loop become infinite? I'm probably just misunderstanding but I'd like to know your thoughts.
